lib/Sema/TypeCheckType.h

```diff
@@ -76,6 +76,12 @@ enum class TypeResolutionFlags : uint16_t {
   /// Pack references are only allowed inside pack expansions
   /// and in generic requirements.
   AllowPackReferences = 1 << 11,
+
+  /// Whether this is a resolution based on a pack expansion.
+  FromPackExpansion = 1 << 12,
```

Comment:
I mentioned this offline, but I wonder if `FromPackExpansion` and `AllowPackReferences` can be one flag, as their uses seem to be very similar

Reply:
Almost same... `maybeResolvePackExpansionType` (where the `AllowPackReferences` flag is added) can be called from `resolvePackExpansionType` (where the `FromPackExpansion` flag is added) but also from `resolveASTFunctionTypeParams`. `AllowPackReferences` is also added in a few other places outside of TypeCheckType.cpp
So running the tests in pack-expansion-expressions.swift, I see cases where `resolvePackReference` is reached and `AllowPackReferences` is set but `FromPackExpansion` is not. However in none of these cases is the new failure case logic in `resolveDeclRefTypeRepr` any different, because in every case of the flag mismatch, either `isParameterPack` and `FromPackReference` are both true or both false (the failure case is only when `isParameterPack` is true but `FromPackReference` is false).
That all said, I don't at this point have the insight into the several other places where the `AllowPackReferences` flag is added to  know how they are reached and whether there could be a scenario where this new diagnostic could give a false negative?

Reply:
Hmmm I wonder if that code in `resolveASTFunctionTypeParams` is now unnecessary / a remnant from when parameter packs used the same syntax as varargs parameters. I commented that code out and am running the tests now locally to see...

The other places where `AllowPackReferences` is set are all during constraint generation when we see an `each T` type repr inside an expression. All of those call-sites look like this:

```cpp
      // Add a PackElementOf constraint for 'each T' type reprs.
      PackExpansionExpr *elementEnv = nullptr;
      if (!PackElementEnvironments.empty()) {
        options |= TypeResolutionFlags::AllowPackReferences;
        elementEnv = PackElementEnvironments.back();
      }
```

and `!PackElementEnvironments.empty()` is a guarantee that the `each` type repr appears inside of a `PackExpansionExpr`, so I don't think these places can cause any false negatives.

Reply:
Oh! There's one other place `AllowPackReferences` is set  in `RequirementRequest` because you can write `each T` in a generic requirement without `repeat`, e.g.

```swift
protocol P {}

func generic<each T>() where each T: P {}
```

...and I think we do want to diagnose something like `where T: P`!

Reply:
Thank you @hborla for the further insights. I have examined and tested more and concluded that it will indeed be best to get rid of my new `FromPackExpansion` flag and just rely on `AllowPackReferences`.
I also found a couple of more tests that needed syntax fixing.